File: TextEditor/src/textEditor/TextEditor.java
package textEditor;


public class TextEditor {
    private String content;
    private int cursorPosition;
    private String selection;

    public TextEditor() {
        this.content = "";
        this.cursorPosition = 0;
        this.selection = "";
    }

    public void setContent(String _content) {
        this.content = _content;
    }

    public void setCursorPosition(int _position) {
        if (_position >= 0 && _position <= content.length())
        {
            this.cursorPosition = _position;
        } else {
            System.out.println("Not valid cursor position.");
        }
    }

    public void addSelection(String _selection) {
        this.selection = _selection;
    }

    public void clearSelections() {
        this.selection = "";
    }

    public void displayState() {
        System.out.println("Content: " + content);
        System.out.println("Cursor Position: " + cursorPosition);
        System.out.println("Selection: " + selection);
    }

    public EditorMemento createMemento(){
        return new EditorMemento(this.content,this.cursorPosition,this.selection);
    }

    public void restoreFromMemento(EditorMemento memento){
        content = memento.getContent();
        cursorPosition = memento.getCursorPosition();
        selection = memento.getSelection();
    }


    public String getContent(){
        return content;
    }

    public int getCursorPosition(){
        return cursorPosition;
    }

    public String getSelection(){
        return selection;
    }
}

File: TextEditor/TextEditor/TextEditor.cs
namespace TextEditor;

public class TextEditor
{
    private string _content;
    private int _cursorPosition;
    private List<string> _selections;

    public TextEditor()
    {
        _content = "";
        _cursorPosition = 0;
        _selections = new List<string>();
    }

    public void SetContent(string content)
    {
        _content = content;
        Console.WriteLine($"Content set to: {_content}");
    }

    public void SetCursorPosition(int position)
    {
        if (position >= 0 && position <= _content.Length)
        {
            _cursorPosition = position;
            Console.WriteLine($"Cursor moved to position: {_cursorPosition}");
        }
        else
        {
            Console.WriteLine("Invalid cursor position.");
        }
    }

    public void AddSelection(string selection)
    {
        _selections.Add(selection);
        Console.WriteLine($"Added selection: {selection}");
    }

    public void ClearSelections()
    {
        _selections.Clear();
        Console.WriteLine("All selections cleared.");
    }

    public void DisplayState()
    {
        Console.WriteLine("\n--- Current Editor State ---");
        Console.WriteLine($"Content: {_content}");
        Console.WriteLine($"Cursor Position: {_cursorPosition}");
        Console.WriteLine("Selections:");
        foreach (var selection in _selections)
        {
            Console.WriteLine($"  - {selection}");
        }
        Console.WriteLine("---------------------------\n");
    }

    public EditorMemento CreateMemento()
    {
        return new EditorMemento(_content, _cursorPosition, new List<string>(_selections));
    }

    public void RestoreFromMemento(EditorMemento memento)
    {
        _content = memento.Content;
        _cursorPosition = memento.CursorPosition;
        _selections = new List<string>(memento.Selections);
        Console.WriteLine(_content);
        Console.WriteLine("Editor state restored from memento.");
    }
}